solutions/gold/izho-17-LongestBeautifulSequence

```diff
@@ -0,0 +1,270 @@
+	for (int i = 0; i < n; i++) {
+		cin >> a[i];
+	}
+	vector<int> k(n);
+	for (int i = 0; i < n; i++) {
+		cin >> k[i];
+	}
+
+	vector<int> dp(n, 1);
+	// best_i: most optimal index to end on
+	int ans = 1, best_i = 0;
+	// prv[i] = optimal index to include right before i
+	vector<int> prv(n);
+	// intially, prv[i] = i
+	iota(prv.begin(), prv.end(), 0);
+
+	for (int i = 1; i < n; i++) {
+		for (int j = 0; j < i; j++) {
+			// bc[a[j]][a[i]] == k[i]
+			if (__builtin_popcount(a[j] & a[i]) == k[i]) {
+				if (dp[j] + 1 > dp[i]) {
+					prv[i] = j;
+					dp[i] = dp[j] + 1;
+				}
+			}
+		}
+		if (dp[i] > ans) {
+			ans = dp[i];
+			best_i = i;
+		}
+	}
+
+	cout << ans << endl;
+
+	vector<int> res;
+	while (prv[best_i] != best_i) {
+		res.push_back(best_i);
+		best_i = prv[best_i];
+	}
+
+	res.push_back(best_i);
+	reverse(res.begin(), res.end());
+	for (int x : res) { cout << x + 1 << " "; }
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
+</Spoiler>
+
+### Subtask 3
+
+While the bounds on $N$ are now much larger ($N \leq 10^5$), the maximum value of $a_i$ is now only $M = 2^8$. We can use this to optimize our solution: rather than looping over every possible previous index, we can loop over all possible previous values instead.
+
+More specifically, we redefine $\texttt{dp}[i]$ to be the length of the longest LBS that ends with **value** $i$ (rather than index). To calculate $\texttt{dp}[a_i]$, we try to transition from $\texttt{dp}[j]$ for all $j$ that satisfy $bc[j][a_i] == k[i]$. This improves our time complexity to $\mathcal{O}(NM)$.
+
+**Time Complexity**: $\mathcal{O}(NM)$
+
+<Spoiler title="Code">
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+#define f first
+#define s second
+using namespace std;
+using pii = pair<int, int>;
+const int B = 8;
+
+int main() {
+    // preprocess:
+	// bc[i][j] = bit_count(i & j)
+	vector<vector<int>> bc(1 << B, vector<int>(1 << B));
+    for (int i = 0; i < (1 << B); i++) {
+        for (int j = 0; j < (1 << B); j++) {
+            bc[i][j] = __builtin_popcount(i & j);
+        }
+    }
+
+	int n;
+    cin >> n;
+    vector<int> a(n);
+    for (int i = 0; i < n; i++) cin >> a[i];
+    vector<int> k(n);
+    for (int i = 0; i < n; i++) cin >> k[i];
+    int ans = 1, best_i = 0; // best_i: best index to end subsequence at
+    vector<int> prv(n); // prv[i]: best index to include before i
+    iota(prv.begin(), prv.end(), 0); // prv[i] = i
+	vector<pii> dp(1 << B, make_pair(-1, -1));
+	// dp[i].f: longest lbs that ends in i
+	// dp[i].s: optimal end index for lbs that ends in i
+
+    for (int i = 0; i < n; i++) {
+		if (dp[a[i]].f == -1) {
+			dp[a[i]].f = 1;
+			dp[a[i]].s = i;
+		}
+        for (int j = 0; j < (1 << B); j++) {
+            if (bc[j][a[i]] == k[i] && dp[j].f + 1 > dp[a[i]].f) {
+				dp[a[i]].f = dp[j].f + 1;
+				dp[a[i]].s = i;
+				prv[i] = dp[j].s;
+			}
+        }
+        if (dp[a[i]].f > ans) {
+            ans = dp[a[i]].f;
+            best_i = i;
```

Comment:
u use spaces here to indent...

Reply:
oh damn that's mb i just pasted it from ide 😅